test/controllers/hosts_controller_test.rb

```diff
@@ -1267,7 +1267,6 @@ class Host::Valid < Host::Managed; end
     setup do
       @host.setBuild
       ActiveRecord::Base.any_instance.expects(:destroy).never
-      ActiveRecord::Base.any_instance.expects(:save).never
```

Comment:
bit unsure how to handle this.

Reply:
Where does the `save` call comes from?

Reply:
app/models/concerns/foreman/sti.rb:26

Reply:
So the two tests that have this issue are https://github.com/theforeman/foreman/blob/develop/test/controllers/hosts_controller_test.rb?utf8=%E2%9C%93#L1324 and https://github.com/theforeman/foreman/blob/develop/test/controllers/hosts_controller_test.rb?utf8=%E2%9C%93#L1336 and it's just because before this change, the `save` call would not be detected, because the actual method was `save_with_type` :D So we were actually calling the save there, and this expectation was not just catching it.

I think the issue was introduced by refactoring in https://github.com/theforeman/foreman/commit/b3fb472c60eeeb085859de58d771a5fec7de7ba3, where the expects were extracted to before filter without realizing there are some tests that actually call save (and we were not hitting issues just because of the aliasing).

We should move the expects just to the tests that really expect nothing to be saved: the explicitness would also help with readability of the expacations IMO

Reply:
Shouldn't the stuff from `describe '#template_used'` be isolated from the `test_submit_multiple_rebuild_config_*` tests?

Reply:
Yes, it seems those two are actually not related to the rest of the context. +1 for getting them out of `template_used`

Reply:
These two are not inside that `describe` block, so I have no idea why the `expects(:save).never` is triggered there

Reply:
ok, I think I got this...